### PR TITLE
dev-java/testng: add 6.11 | java-utils-2.eclass: allow etestng to run testng.xml

### PR DESCRIPTION
**Pull request by vaukai**

### CODE REVIEW

**Comment by vaukai** [comment]
@fordfrog 
As the next of my open PRs please merge this one.
5 commits where "avoid reverse dependency" is split by eclass.
I selected version `6.11`  according to vulnerabilities listed on [mvnrepository](https://mvnrepository.com/artifact/org.testng/testng) (I have version `7.5.1` WIP, but still fighting with some test failures).

Another, hopefully less time consuming, ready to merge is https://github.com/gentoo/gentoo/pull/32973.

**Comment by fordfrog** [comment]
i get this warning during the test phase:
```
 * java-pkg_ensure-dep: The ebuild is attempting to use guava, which is not declared with a SLOT in DEPEND.
```
and then later this failure:
```
[TestNG] [ERROR] 
Cannot find class in classpath: test.groovy.GroovyTest
```

**Comment by vaukai** [comment]
> i get this warning during the test phase:
> 
> ```
>  * java-pkg_ensure-dep: The ebuild is attempting to use guava, which is not declared with a SLOT in DEPEND.
> ```

@fordfrog I've now added guava to DEPEND. 

> and then later this failure:
> 
> ```
> [TestNG] [ERROR] 
> Cannot find class in classpath: test.groovy.GroovyTest
> ```

This ERROR, I cannot even reproduce. I always get that class as
```
target/test-classes/test/groovy/GroovyTest.class
```
and `target/test-classes` is always on classpath.
Could you pls check if `GroovyTest.class` is there? And can you reproduce that error is you ran the tests 3 ofr 4 times?

**Comment by fordfrog** [comment]
now i get this warning in install phase:
```
 * java-pkg_ensure-dep: The ebuild is attempting to use testng, which is not declared with a SLOT in [RP]DEPEND and --build-only wasn't given.
```
i'm not sure the system package should be used in testng when you wanna use the internal one, if i get it right.

searching for `GroovyTest.class` gives me nothing:
```
find /var/tmp/portage/dev-java/testng-6.11/work/ -name GroovyTest.class
```
here's the full [build.log](https://github.com/gentoo/gentoo/files/12776641/build.log)

**Comment by vaukai** [comment]
> now i get this warning in install phase:
> 
> ```
>  * java-pkg_ensure-dep: The ebuild is attempting to use testng, which is not declared with a SLOT in [RP]DEPEND and --build-only wasn't given.
> ```
> 
> i'm not sure the system package should be used in testng when you wanna use the internal one, if i get it right.

Would you agree to suppress that warning? When just building and testing `testng` it's not yet installed, so we cannot have it in [RP]DEPEND.

**Comment by vaukai** [comment]
> searching for `GroovyTest.class` gives me nothing:
> 
> ```
> find /var/tmp/portage/dev-java/testng-6.11/work/ -name GroovyTest.class
> ```

I have to investigate ...

**Comment by fordfrog** [comment]
> > now i get this warning in install phase:
> > ```
> >  * java-pkg_ensure-dep: The ebuild is attempting to use testng, which is not declared with a SLOT in [RP]DEPEND and --build-only wasn't given.
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > i'm not sure the system package should be used in testng when you wanna use the internal one, if i get it right.
> 
> Would you agree to suppress that warning? When just building and testing `testng` it's not yet installed, so we cannot have it in [RP]DEPEND.

which line does this cause exactly?

**Comment by vaukai** [comment]
> which line does this cause exactly?

It's in the following two commits, both titeled `"avoid reverse dependency for dev-java/testng"`

**Comment by vaukai** [comment]
@fordfrog 
thanks for trying again. I still don't have any idea why it fails for you but works here. Will need some time to investigate.

**Comment by vaukai** [comment]
@fordfrog 
From my point of view it's now ready to merge. 
Let's handle the `problem with multi-jar packages cumulating tests` separately. Could we do something with `unset <valiable>` (I have no experience with)?

**Comment by vaukai** [comment]
> still failing for me: [build.log](https://github.com/gentoo/gentoo/files/12828058/build.log)

@fordfrog 
The problem is that for multi-jar packages we have and need in `java-pkg-simple.eclass`
```
    # https://bugs.gentoo.org/906311
    rm -rf ${classes} || die
```
But this also removes all `test-classes` generated before calling [java-pkg-simple_src_test()](https://github.com/gentoo/gentoo/pull/29733/commits/1e4bc2ddaa56d81a4eb16e5f3235f8ebd7de898e#diff-435ce2cb04d436e92c207ad4dd68dc3301c7fbd63edc602f23d1527b7c19e036R87) in an `ebuild`.

Guess we need to move generated `test-classes` somewhere else and add them back after `rm -rf ${classes}`?

**Comment by vaukai** [comment]
Problem solved

**Comment by vaukai** [comment]
> > > now i get this warning in install phase:
> > > ```
> > >  * java-pkg_ensure-dep: The ebuild is attempting to use testng, which is not declared with a SLOT in [RP]DEPEND and --build-only wasn't given.
> > > ```
> > > 
> > > 
> > > i'm not sure the system package should be used in testng when you wanna use the internal one, if i get it right.
> > 
> > 
> > Would you agree to suppress that warning? When just building and testing `testng` it's not yet installed, so we cannot have it in [RP]DEPEND.
> 
> which line does this cause exactly?

@fordfrog 
That warning would go away if that line with `java-pkg_register-dependency testng jquery.jar` was commented out. But `jquery.jar` is needed at runtime.  Do you have a better idea to get rid of that warning?

**Comment by fordfrog** [comment]
> @fordfrog That warning would go away if that line with `java-pkg_register-dependency testng jquery.jar` was commented out. But `jquery.jar` is needed at runtime. Do you have a better idea to get rid of that warning?

isn't `java-pkg_regjar` more appropriate in this case?

File: eclass/java-pkg-simple.eclass [hunk 1]
```diff
@@ -506,6 +507,11 @@ java-pkg-simple_src_test() {
 	# create the target directory
 	mkdir -p ${classes} || die "Could not create target directory for testing"
 
+	# generated test classes should get generated into "generated-test" directory
+	if [[ -d generated-test ]]; then
+		cp -r generated-test/* "${classes}" || die "cannot copy generated test classes"
+	fi
+
```
Comment (fordfrog) [inline eclass/java-pkg-simple.eclass:514]: i think this feature should be documented in the function description

**Comment by fordfrog** [comment]
i merged it as it is, just please see my note with regard to `generated-test`

File: eclass/java-pkg-simple.eclass (hunk 1 above)
Comment (vaukai) [reply eclass/java-pkg-simple.eclass:514]: Just thinking about a good explanation ...

**Pull request closed without merging**